### DISCUSSION TITLE
[completed] stage 2 - Upload and CRUD forms

### DISCUSSION THREAD

**Discussion by atherdon** [atherdon/dash-back#42] [General]
@kolserdav предлагаю обсудить майлстоун для upload данных и react app для добавления этих данных в базу.

Думаю auth-ом просто нас пытаются отвлечь - эту проблему всегда можно будет решить позже - сейчас это не критично.

а вот момент переноса релевантных данных в нашу систему важен и если мы сделаем его - это будет экономить время on-daily-basis помощницы редактора. поидее тут нет ничего сложного, нужно просто обсудить какие схемы есть сейчас, какие схемы нужно добавить в backend и как потом сохранять это в sqllite.

Я могу снять видео, и показать процесс того, как помощница сейчас работает. наверное это нужно, для понимания всех деталей

**Comment by kolserdav** [comment]
Как работает помощница и без видео понятно) лучше покажите схемы таблиц, которые нужно перенести. Там `mongoDb`,? Файлы в  `data` это вывод из той базы?

**Reply by atherdon** [reply]
файлы в data - это просто экспорт из excel.

**Reply by atherdon** [reply]
доступа к освной базе нет и не обязательно сейчас.

**Reply by kolserdav** [reply]
Из exel надо в базу залить?

**Reply by kolserdav** [reply]
Если из базы, тогда нужно знать схемы таблиц и какая база

**Reply by atherdon** [reply]
сейчас запишу видео

**Reply by atherdon** [reply]
_(у меня телеграм лагает, поэтому дублирую тут)_

> как я понял нужно переименовать brands в everygreen, а также сделать скрипт который зальет из brands.js editors.js и top-authors.js данные в соответствующие таблицы базы. И добавить CRUD на ещё какие-то схемы и их тоже залить из data?

ну первую часть мы обсуждали в этом таске 
https://github.com/atherdon/dash-back/issues/18

- нужен скрипт для импорта в текущие схемы
- можно сделать схемы и для других файлов из data(рано или поздно все равно это нужно будет сделать)
- создать схему для Articles(просто клон brands или topAuthors)
- React app c формами для наших схем для добавления данных в нашу базу. я думал найти автогенератор для этих форм. чтобы мы могли скормить схему из graphql и реакт бы автоматически их генерировал, чтобы не делать с нуля

сс @kolserdav

**Reply by kolserdav** [reply]
По первым трем пунктам всё понятно. 4 пункт тоже понятно, но у меня возникли проблемки с отловом ошибок на `react-apollo-form`, поразбираюсь, возможно немного бек надо будет под него подкрутить.

**Reply by kolserdav** [reply]
что-то с `react-apollo-form` не так( у вас нет позитивных практик в этом направлении, что лучше использовать?

**Reply by kolserdav** [reply]
так-то он формы создает и довольно быстро его юзать, но он ориентируется только по типам данных, и если произошла ошибка не валидности данных которые проходят по типу, то у него нет никаких инструментов их обработать.

**Reply by kolserdav** [reply]
есть ещё `react-apollo-forms`, там `onError` есть, но судя по количеству загрузок, не похоже это на хорошее решение как и `react-apollo-form`

**Reply by atherdon** [reply]
я только слышал о `react-apollo-form`. просто на php когда я работал у нас были автоматические формы из конфига моделей, поэтому я подумал что и для js такое должно быть.

вот что я нагуглил, но нужно пробывать.

1. https://www.youtube.com/watch?v=Ovg9CYwWFBM&ab_channel=ReactEurope
https://github.com/rjsf-team/react-jsonschema-form

2. https://blog.usejournal.com/build-react-forms-based-on-graphql-apis-7e3acb1f91d7
3. https://blog.logrocket.com/quickly-build-schema-based-forms-in-react-with-uniforms/
https://uniforms.tools/ 
https://github.com/vazco/uniforms

4. https://www.npmjs.com/package/@ttoohey/react-form

сс @kolserdav

**Reply by kolserdav** [reply]
Я тоже гуглил и `react-apollo-form` мне показался наиболее подходящим вариантом, который позволяет быстрее всего встроиться в наш бек. Но он не подходит из-за невозможности обработки нештатных ситуаций, пробовал форкнуть его и доработать, но там даже проект не получается собрать (в `scripts` ничего нет, а просто `tsc -p .` вываливает ошибки). Для временного решения, чтобы быстро испытывать бек - это идеальное решение, но для пользователя я бы его не стал использовать, так как если что-то пойдет не так, то приложение не сможет понятным для пеользователя языком об этом сообщить. Так как при любой ошибке сервера на деве он просто выбразывает исключение на весь экран, а на деплое делает вид что ничего не произошло.
Поэтому по последнему пункту могу лишь сделать  на `@apollo/client`.  Там благодаря `ApolloProvider` сможем юзать хук `useMutation`, что позволит выводить ошибки сервера в форму. Но `useMutation` слишком высокоуровневый для разделения ошибок на статусы `error` | `warning`, так как выдает в результаты только сообщение `message` об ошибке, игнорируя специальные поля ошибок которые может возвращать бекенд, то есть например: ошибка сервера "ошибка создания хеша пароля" и например "емайл уже зарегистрирован" для него будут выглядеть одинаково (красным) только разные сообщения - но это хотя бы решает проблему `react-apollo-form`, который вообще не давал никак обрабатывать нештатные ситуации. Если важно, чтобы вывод сервера отличался по цвету, когда `warning` желтый например, а когда `error` то красный, то напишите, попробую без `useMutation`.

**Reply by atherdon** [reply]
@kolserdav для меня - crud формы для наших схем -  это всего лиш временный вариант, который поможет сейчас не делать sync с основной базой. после вашего сообщения - я думаю что мы можем использовать react-apollo-form.

если мы сделаем backend правильно(что мы и пытаемся сейчас) - то я думаю Марина будет использовать эти crud формы максимум месяца 2 после запуска. скажите тогда эстимейт и я создам проект на FH

**Reply by kolserdav** [reply]
> если мы сделаем backend правильно(что мы и пытаемся сейчас)

что Вы имеете в виду?

**Reply by atherdon** [reply]
:) я про то что если мы также будем двигаться дальше - то примерно через месяц мы запустим уже редакторов внутрь проекта. просто я еще пытаюсь предусмотреть вдруг чтото пойдет нетак

**Reply by kolserdav** [reply]
Тогда день понадобится на добавление остальных моделей и создание скриптов для заливки их в базу и день для реакт апп с формами для наполнения. У нас только регистрация имеет проверки, кроме проверок типов, поэтому её сделаю на `useMutation` а остальные на `react-apollo-form`. За два дня планирую управиться.

**Reply by atherdon** [reply]
Отлично - создаю тогда проект на frelancehunt

**Reply by kolserdav** [reply]
принял

**Reply by atherdon** [reply]
отправил оплату на fh - ждем когда обработается.
статус тут: https://www.coinpayments.net/index.php?cmd=status&id=CPFB5GTXSLZUFVFCZEXSEH1IIK&key=1d58e19952c138b9cfdc7f8e45189099
